File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/disconnectfour/domain/DisconnectFourGameState.java
package com.zwstudio.logicpuzzlesandroid.puzzles.disconnectfour.domain;

import com.zwstudio.logicpuzzlesandroid.common.domain.AllowedObjectState;
import com.zwstudio.logicpuzzlesandroid.common.domain.CellsGameState;
import com.zwstudio.logicpuzzlesandroid.common.domain.Position;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import fj.function.Effect0;

/**
 * Created by zwvista on 2016/09/29.
 */

public class DisconnectFourGameState extends CellsGameState<DisconnectFourGame, DisconnectFourGameMove, DisconnectFourGameState> {
    public DisconnectFourObject[] objArray;
    public Map<Position, AllowedObjectState> pos2state = new HashMap<>();

    public DisconnectFourGameState(DisconnectFourGame game) {
        super(game);
        objArray = new DisconnectFourObject[rows() * cols()];
        System.arraycopy(game.objArray, 0, objArray, 0, objArray.length);
        updateIsSolved();
    }

    public DisconnectFourObject get(int row, int col) {
        return objArray[row * cols() + col];
    }
    public DisconnectFourObject get(Position p) {
        return get(p.row, p.col);
    }
    public void set(int row, int col, DisconnectFourObject dotObj) {
        objArray[row * cols() + col] = dotObj;
    }
    public void set(Position p, DisconnectFourObject obj) {
        set(p.row, p.col, obj);
    }

    public boolean setObject(DisconnectFourGameMove move) {
        if (!isValid(move.p) || game.get(move.p) != DisconnectFourObject.Empty || get(move.p).equals(move.obj)) return false;
        set(move.p, move.obj);
        updateIsSolved();
        return true;
    }

    public boolean switchObject(DisconnectFourGameMove move) {
        if (!isValid(move.p) || game.get(move.p) != DisconnectFourObject.Empty) return false;
        DisconnectFourObject o = get(move.p);
        move.obj = o == DisconnectFourObject.Empty ? DisconnectFourObject.Yellow :
            o == DisconnectFourObject.Yellow ? DisconnectFourObject.Red:
            DisconnectFourObject.Empty;
        return setObject(move);
    }

    /*
        iOS Game: Logic Games/Puzzle Set 11/Disconnect Four

        Summary
        Win by not winning!

        Description
        1. The opposite of the famous game 'Connect Four', where you must line
           up four tokens of the same colour.
        2. In this puzzle you have to ensure that there are no more than three
           tokens of the same colour lined up horizontally, vertically or
           diagonally.
    */
    private void updateIsSolved() {
        isSolved = true;
        for (int r = 0; r < rows(); r++)
            for (int c = 0; c < cols(); c++) {
                Position p = new Position(r, c);
                pos2state.put(p, AllowedObjectState.Normal);
            }
        DisconnectFourObject oLast = DisconnectFourObject.Empty;
        List<Position> trees = new ArrayList<>();
        Effect0 checkTrees = () -> {
            if (trees.size() > 3) {
                isSolved = false;
                for (Position p : trees)
                    pos2state.put(p, AllowedObjectState.Error);
            }
            trees.clear();
        };
        for (int r = 0; r < rows(); r++) {
            oLast = DisconnectFourObject.Empty;
            for (int c = 0; c < cols(); c++) {
                Position p = new Position(r, c);
                DisconnectFourObject o = get(p);
                if (o != oLast) {
                    checkTrees.f();
                    oLast = o;
                }
                if (o == DisconnectFourObject.Empty)
                    isSolved = false;
                else
                    trees.add(p);
            }
            checkTrees.f();
        }
        for (int c = 0; c < cols(); c++) {
            oLast = DisconnectFourObject.Empty;
            for (int r = 0; r < rows(); r++) {
                Position p = new Position(r, c);
                DisconnectFourObject o = get(p);
                if (o != oLast) {
                    checkTrees.f();
                    oLast = o;
                }
                if (o == DisconnectFourObject.Empty)
                    isSolved = false;
                else
                    trees.add(p);
            }
            checkTrees.f();
        }
    }
}

File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/disconnectfour/domain/DisconnectFourGameState.kt
package com.zwstudio.logicpuzzlesandroid.puzzles.disconnectfour.domain

import com.zwstudio.logicpuzzlesandroid.common.domain.AllowedObjectState
import com.zwstudio.logicpuzzlesandroid.common.domain.CellsGameState
import com.zwstudio.logicpuzzlesandroid.common.domain.Position

class DisconnectFourGameState(game: DisconnectFourGame) : CellsGameState<DisconnectFourGame, DisconnectFourGameMove, DisconnectFourGameState>(game) {
    var objArray = game.objArray.copyOf()
    var pos2state = mutableMapOf<Position, AllowedObjectState>()

    operator fun get(row: Int, col: Int) = objArray[row * cols() + col]
    operator fun get(p: Position) = this[p.row, p.col]
    operator fun set(row: Int, col: Int, dotObj: DisconnectFourObject) {objArray[row * cols() + col] = dotObj}
    operator fun set(p: Position, obj: DisconnectFourObject) {this[p.row, p.col] = obj}

    init {
        updateIsSolved()
    }

    fun setObject(move: DisconnectFourGameMove): Boolean {
        val p = move.p
        if (!isValid(p) || game[p] != DisconnectFourObject.Empty || this[p] == move.obj) return false
        this[p] = move.obj
        updateIsSolved()
        return true
    }

    fun switchObject(move: DisconnectFourGameMove): Boolean {
        val p = move.p
        if (!isValid(p) || game[p] != DisconnectFourObject.Empty) return false
        val o = this[p]
        move.obj = if (o == DisconnectFourObject.Empty) DisconnectFourObject.Yellow else if (o == DisconnectFourObject.Yellow) DisconnectFourObject.Red else DisconnectFourObject.Empty
        return setObject(move)
    }

    /*
        iOS Game: Logic Games/Puzzle Set 11/Disconnect Four

        Summary
        Win by not winning!

        Description
        1. The opposite of the famous game 'Connect Four', where you must line
           up four tokens of the same colour.
        2. In this puzzle you have to ensure that there are no more than three
           tokens of the same colour lined up horizontally, vertically or
           diagonally.
    */
    private fun updateIsSolved() {
        isSolved = true
        for (r in 0 until rows())
            for (c in 0 until cols()) {
                val p = Position(r, c)
                pos2state[p] = AllowedObjectState.Normal
            }
        var oLast: DisconnectFourObject
        val trees = mutableListOf<Position>()
        fun checkTrees() {
            if (trees.size > 3) {
                isSolved = false
                for (p in trees)
                    pos2state[p] = AllowedObjectState.Error
            }
            trees.clear()
        }
        for (r in 0 until rows()) {
            oLast = DisconnectFourObject.Empty
            for (c in 0 until cols()) {
                val p = Position(r, c)
                val o = this[p]
                if (o != oLast) {
                    checkTrees()
                    oLast = o
                }
                if (o == DisconnectFourObject.Empty)
                    isSolved = false
                else
                    trees.add(p)
            }
            checkTrees()
        }
        for (c in 0 until cols()) {
            oLast = DisconnectFourObject.Empty
            for (r in 0 until rows()) {
                val p = Position(r, c)
                val o = this[p]
                if (o != oLast) {
                    checkTrees()
                    oLast = o
                }
                if (o == DisconnectFourObject.Empty)
                    isSolved = false
                else
                    trees.add(p)
            }
            checkTrees()
        }
    }
}